src/hotspot/share/opto/vmaskloop.hpp

```diff
@@ -0,0 +1,136 @@
+class VectorMaskedLoop : public ResourceObj {
+ private:
+  // Useful handles
+  PhaseIdealLoop*            _phase;
+  PhaseIterGVN*              _igvn;
+  Arena*                     _arena;
+
+  // Loop information
+  IdealLoopTree*             _lpt;          // Idealloop tree
+  CountedLoopNode*           _cl;           // CountedLoop node
+  CountedLoopEndNode*        _cle;          // CountedLoopEnd node
+  PhiNode*                   _iv;           // Loop induction variable PhiNode
+
+  // Data structures for loop analysis
+  Unique_Node_List           _core_set;     // Loop core nodes set for fast membership check
+  Unique_Node_List           _body_set;     // Loop body nodes set for fast membership check
+  GrowableArray<Node*>       _body_nodes;   // Loop body nodes with reverse postorder
+  GrowableArray<int>         _rpo_idx;      // Map from node index to RPO traversal index
+  GrowableArray<BasicType>   _elem_bt;      // Per node vector element basic type
+  GrowableArray<Node_List*>  _stmts;        // Lists of nodes that make up loop statements
+  GrowableArray<SWPointer*>  _swptrs;       // SWPointer array for memory access nodes
+  VectorElementSizeStats     _size_stats;   // Statistics of data sizes in vectors
+
+  // Basic utilities
+  bool in_core(Node* n)            { return n != nullptr && _core_set.member(n); }
+  bool in_body(Node* n)            { return n != nullptr && _body_set.member(n); }
+  int  rpo_idx(Node* n)            { assert(in_body(n), "What?"); return _rpo_idx.at(n->_idx); }
+  void set_rpo_idx(Node* n, int i) { assert(in_body(n), "What?"); _rpo_idx.at_put_grow(n->_idx, i); }
+
+  BasicType statement_bottom_type(Node_List* stmt) {
+    assert(stmt != nullptr && stmt->size() > 0, "should not be empty");
+    assert(stmt->at(0)->is_Store(), "Must be a store node");
+    return stmt->at(0)->as_Store()->memory_type();
+  }
+
+  BasicType size_to_basic_type(int size) {
+    BasicType bt = T_ILLEGAL;
+    switch (size) {
+      case 1: bt = T_BYTE;  break;
+      case 2: bt = T_SHORT; break;
+      case 4: bt = T_INT;   break;
+      case 8: bt = T_LONG;  break;
+      default: ShouldNotReachHere();
+    }
+    return bt;
+  }
+
+  // Node vector element type accessors
+  BasicType elem_bt(Node* n) { return _elem_bt.at(rpo_idx(n)); }
+  void set_elem_bt(Node* n, BasicType bt) { _elem_bt.at_put(rpo_idx(n), bt); }
+  bool has_valid_elem_bt(Node* n) { return elem_bt(n) != T_ILLEGAL; }
+
+  // Some node check utilities
+  bool is_loop_iv(Node* n) { return n == _iv; }
+  bool is_loop_incr(Node* n) { return n == _cl->incr(); }
+  bool is_loop_iv_or_incr(Node* n) { return n == _iv || n == _cl->incr(); }
+
+  bool is_loop_iv_plus_stride(Node* n) {
+    if (n != nullptr && n->is_Add() && n->in(1) == _iv && n->in(2)->is_Con()) {
+      const Type* t = n->in(2)->bottom_type();
+      return t->is_int()->get_con() == _cl->stride_con();
+    }
+    return false;
+  }
```

Comment:
Do you not want to do this sort of implementation in `SWPointer` instead? There are already methods like `scaled_iv_plus_offset`, so it would fit in next to that, right?

Reply:
It doesn't fit well as functions in `SWPointer` can only be used for checking the pattern in indices. But this function may be used for checking the loop increment pattern which is not in array indices, perhaps `a[i] = b[i] * (i + 1)`. We don't have `SWPointer` constructed for this. I have rename the function to make the purpose clear.

Reply:
Ah, you are right, that is a different use. Yes, better function name often does the trick :blush: